src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSetSerializer.cs

```diff
@@ -357,11 +357,30 @@ private IEnumerable<ODataOperation> CreateODataOperations(IEnumerable<IEdmOperat
         private static Uri GetNestedNextPageLink(ODataSerializerContext writeContext, int pageSize, object obj)
         {
             Contract.Assert(writeContext.ExpandedResource != null);
+            Uri nestedNextLink = null;
             IEdmNavigationSource sourceNavigationSource = writeContext.ExpandedResource.NavigationSource;
-            NavigationSourceLinkBuilderAnnotation linkBuilder = writeContext.Model.GetNavigationSourceLinkBuilder(sourceNavigationSource);
-            Uri navigationLink =
-                linkBuilder.BuildNavigationLink(writeContext.ExpandedResource, writeContext.NavigationProperty);
-            Uri nestedNextLink = GenerateQueryFromExpandedItem(writeContext, navigationLink);
+
+            // In Contained Navigation, we don't have navigation property binding,
+            // Hence we cannot get the NavigationLink from the NavigationLinkBuilder
+            if (writeContext.NavigationSource.NavigationSourceKind() == EdmNavigationSourceKind.ContainedEntitySet)
+            {
+                // Contained navigation.
+                var linkBuilder = new NavigationSourceLinkBuilderAnnotation(sourceNavigationSource, writeContext.Model);
+                var idlink = linkBuilder.BuildIdLink(writeContext.ExpandedResource);
+
+                var link = idlink.ToString() + "/" + writeContext.NavigationSource.Name;
```

Comment:
> var link = idlink.ToString() + "/" + writeContext.NavigationSource.Name; [](start = 15, length = 73)

I can't understand here.

Id link itself has the navigation source, so, append the navigation source again? 

Can you add the test to cover the output link?

Reply:
I think that the NavigationSourceLinkBuilderAnnotation.BuildIdLink builds the Uri for a particular contained resource (i.e., writeContext.ExpandedResource).  We then need to append the name of the navigation property on that contained resource.

---
In reply to: [488983874](https://github.com/OData/WebApi/pull/2287#discussion_r488983874) [](ancestors = 488983874)

Reply:
@xuzhg In a scenario where we want to build the `NavigationLink` as `https://baseUri/Stores(1)/Employees`,
BuildIdLink will return the Uri for the contained resource as `https://baseUri/Stores(1)`. Therefore we need to append the navigation property `Employees`.

Reply:
@KenitoInc I mean why append the navigation source name, it should be the "navigation property name", right?

Reply:
i.e.; should line 371 be wirteContext.NavigationProperty.Name instead of writeContext.NavigationSource.Name?

---
In reply to: [489627956](https://github.com/OData/WebApi/pull/2287#discussion_r489627956) [](ancestors = 489627956)